Feature/soft ask photo picker

nbradbury opened the pull request:
Introduces "soft asks" for permissions in the photo picker, which is used in the editor and the gravatar selection. Note that this moves permission logic out of those areas - it's handled entirely by the picker now.

When the picker is shown and the user hasn't granted the app the required permissions, instead of displaying a grid of photos it will display a "soft ask" message that it needs permission to access photos.

![permission-needed](https://user-images.githubusercontent.com/3903757/27174932-b401e1ac-518b-11e7-8df7-17b7fa911f0f.png)

Tapping the Allow button shows the native permission dialog.

![permissions-allow](https://user-images.githubusercontent.com/3903757/27174946-bbc34106-518b-11e7-91a6-078ae55bde7b.png)

If the user allows the permissions, the picker will display photos. If permissions are denied, the soft ask will continue to appear. If the permissions are denied and the user checked "Never ask again," the Allow button changes to Edit Permissions, and tapping that takes the user to the app settings so they can manually edit the permissions.

![permissions-denied](https://user-images.githubusercontent.com/3903757/27174958-c33520b2-518b-11e7-8df7-5d3688861182.png)

daniloercoli commented on WordPress/src/main/java/org/wordpress/android/ui/photopicker/PhotoPickerFragment.java:
```diff
@@ -74,49 +85,48 @@
-        mPhotosOnly = options != null && options.contains(PhotoPickerOption.PHOTOS_ONLY);
+    public void setOptions(@NonNull EnumSet<PhotoPickerOption> options) {
+        mAllowMultiSelect = options.contains(PhotoPickerOption.ALLOW_MULTI_SELECT);
+        mPhotosOnly = options.contains(PhotoPickerOption.PHOTOS_ONLY);
```
I might be wrong, but it seems that `mDeviceOnly` is missing here. 

Minor: Wondering why there is this method to refresh/set Fragment parameters. Is there a way to switch the picker mode after its creation? I see it's used by the main activity in `onCreate`, when the fragment is already available.

daniloercoli commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -1676,6 +1676,10 @@
     <string name="photo_picker_capture_video">Take video</string>
+    <string name="photo_picker_soft_ask_label">%s needs permissions to access your photos</string>
+    <string name="photo_picker_soft_ask_allow">Allow</string>
+    <string name="photo_picker_soft_ask_permissions_denied">%s was denied permissions to access your photos</string>
```
Should we change this to `...to access your media`?

nbradbury replied:
> Should we change this to ...to access your media?

I went back and forth about this. I figured most people are adding pictures so using "photos" made the most sense, but it isn't really accurate since it shows videos as well. What do you think?

daniloercoli replied:
> I went back and forth about this. I figured most people are adding pictures so using "photos" made the most sense, but it isn't really accurate since it shows videos as well. What do you think?

Both ways are fine to me. Just noticed that we've used the wording `optimize images` in settings screen, in section about compressing/resizing images. Worth to change both to `photos`? 

Also noticed that the system dialog, shown when I clicked on Allow, has the following wording `Allow <app-name> to access photos, media, and files on your devices`.

daniloercoli replied:
As per chat convo, we've agreed that `Photos` is the better choice in this context.

nbradbury commented:
> I might be wrong, but it seems that mDeviceOnly is missing here.

Thanks for catching that! Fixed in 1df7faa.

> Wondering why there is this method to refresh/set Fragment parameters. Is there a way to switch the picker mode after its creation?

It probably isn't necessary, but I added it since the PhotoPicker was built to be reused.

daniloercoli commented:
Looks good! :shipit:
